aptos-move/framework/aptos-framework/sources/transaction_validation.move

```diff
@@ -70,6 +72,25 @@ module aptos_framework::transaction_validation {
         });
     }
 
+    /// Called in prologue to optionally hold some amount for special txns (e.g. randomness txns).
+    /// `return_deposit()` should be invoked in the corresponding epilogue with the same arguments.
+    fun collect_deposit(gas_payer: address, amount: Option<u64>) {
+        if (option::is_some(&amount)) {
+            let amount = option::extract(&mut amount);
+            let balance = coin::balance<AptosCoin>(gas_payer);
+            assert!(balance >= amount, error::invalid_state(PROLOGUE_EINSUFFICIENT_BALANCE_FOR_REQUIRED_DEPOSIT));
+            transaction_fee::burn_fee(gas_payer, amount);
+        }
+    }
+
+    /// Called in epilogue to optionally released the amount held in prologue for special txns (e.g. randomness txns).
+    fun return_deposit(gas_payer: address, amount: Option<u64>) {
+        if (option::is_some(&amount)) {
+            let amount = option::extract(&mut amount);
+            transaction_fee::mint_and_refund(gas_payer, amount);
+        }
+    }
+
```

Comment:
@movekevin @gedigi Do you guys feel safe about burning and then re-minting the locked deposit? I fear that in the case of a VM bug, we may end up minting new coins. I wonder if it would be better to implement some sort of actual lock-up.

Reply:
we already do this for storage refunds, so although conceptually not ideal it doesn't seem to have introduced consequences. minting and burning doesn't generate events - so event consumers shouldn't take unexpected actions. and the total supply is an aggregator, so it shouldn't interfere with parallelism.

Reply:
@igor-aptos was also suggesting we return the amount as a coin object to rust in prologue and pass it back to move in epilogue, to avoid burning/minting, which sounds ideal if it works. I'm gonna verify it.

Reply:
> we return the amount as a coin object to rust in prologue and pass it back to move in epilogue

This might make it slightly safer but doesn't address the issue fundamentally, as the Move VM isn't bound to to honor any rules around resource safety -- it could still in theory treat this object as raw bytes and make unintended modifications.